File: dynamic_programming/house_robber_198.go
package dynamic_programming

func rob(nums []int) int {
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	n += 1
	dp := make([]int, n)
	dp[0] = nums[0]
	dp[1] = maxRob(nums[0], nums[1])
	for i := 2; i < n; i++ {
		f := dp[i-1]
		s := dp[i-2]
		if i < len(nums) {
			s += nums[i]
		}
		dp[i] = maxRob(f, s)
	}
	return dp[n-1]
}

func maxRob(f, s int) int {
	if f > s {
		return f
	}
	return s
}

File: dynamic_programming/java_solutions/HouseRobber_198.java
package dynamic_programming.java_solutions;

public class HouseRobber_198 {

    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 1) {
            return nums[0];
        }
        int[] dp = new int[n];

        dp[0] = nums[0];
        dp[1] = Math.max(nums[1], nums[0]);

        for (int i = 2; i < n; i++) {
            dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1]);
        }
        return dp[n - 1];
    }
}
